tasty-query/shared/src/main/scala/tastyquery/Annotations.scala

```diff
@@ -1,9 +1,106 @@
 package tastyquery
 
+import scala.annotation.tailrec
+
+import tastyquery.Constants.*
+import tastyquery.Contexts.*
+import tastyquery.Exceptions.*
+import tastyquery.Names.*
+import tastyquery.Symbols.*
 import tastyquery.Trees.*
+import tastyquery.Types.*
 
 object Annotations:
   final class Annotation(val tree: TermTree):
+    private var mySymbol: ClassSymbol | Null = null
+    private var myArguments: List[TermTree] | Null = null
+
+    /** The annotation class symbol. */
+    def symbol(using Context): ClassSymbol =
+      val local = mySymbol
+      if local != null then local
+      else
+        val computed = computeAnnotSymbol(tree)
+        mySymbol = computed
+        computed
+    end symbol
+
+    /** The symbol of the constructor used in the annotation. */
+    def annotConstructor(using Context): TermSymbol =
+      computeAnnotConstructor(tree)
+
+    /** All the term arguments to the annotation's constructor.
+      *
+      * If the constructor has several parameter lists, the arguments are
+      * flattened in a single list.
+      *
+      * `NamedArg`s are not visible with this method. They are replaced by
+      * their right-hand-side.
+      */
+    def arguments(using Context): List[TermTree] =
+      val local = myArguments
+      if local != null then local
+      else
+        val computed = computeAnnotArguments(tree)
+        myArguments = computed
+        computed
+    end arguments
+
+    def argCount(using Context): Int = arguments.size
+
+    def argIfConstant(idx: Int)(using Context): Option[Constant] =
+      arguments(idx) match
+        case Literal(constant) => Some(constant)
+        case _                 => None
+
     override def toString(): String = s"Annotation($tree)"
   end Annotation
+
+  private def computeAnnotSymbol(tree: TermTree)(using Context): ClassSymbol =
+    def invalid(): Nothing =
+      throw InvalidProgramStructureException(s"Cannot find annotation class in $tree")
+
+    @tailrec
+    def loop(tree: TermTree): ClassSymbol = tree match
+      case Apply(fun, _)       => loop(fun)
+      case Select(New(tpt), _) => tpt.toType.classSymbol.getOrElse(invalid())
+      case TypeApply(fun, _)   => loop(fun)
+      case Block(_, expr)      => loop(expr)
+      case _                   => invalid()
```

Comment:
there's some ancient tasty out there with raw New as its tree, btw

Reply:
the code gen for enums at some point had a raw `New(unchecked)` annotation

Reply:
OK I adapted `computeAnnotSymbol` and `computeAnnotArguments` to handle that case. I left `computeAnnotConstructor` unchanged, as in that case we should really blow up I think.